tests/test.FsWatcher.js

```diff
@@ -0,0 +1,177 @@
+ */
+
+/*
+ * Copyright (c) 2015, Joyent, Inc.
+ */
+
+var execFile = require('child_process').execFile;
+
+var test = require('tape');
+var vmadm = require('vmadm');
+
+var common = require('./common');
+var mocks = require('./mocks');
+var FsWatcher = require('../lib/watchers/fs-watcher');
+
+
+// How frequently to poll the 'events' array when we're waiting for an event.
+var EVENTS_POLL_FREQ = 100; // ms
+
+var events = [];
+var existingVms = [];
+var smartosImageUUID;
+var smartosVmUUID;
+var watcher;
+
+
+function waitEvent(t, evt, vmUuid, eventIdx) {
+    var loops = 0;
+
+    function _waitEvent() {
+        var i;
+
+        if (events.length > eventIdx) {
+            // we've had some new events, check for our create
+            for (i = eventIdx; i < events.length; i++) {
+                if (events[i].vmUuid === vmUuid && events[i].event === evt) {
+                    t.ok(true, 'FsWatcher saw expected ' + evt
+                        + ' (' + (loops * EVENTS_POLL_FREQ) + ' ms)');
+                    t.end();
+                    return;
+                }
+            }
+        }
+
+        loops++;
+        setTimeout(_waitEvent, EVENTS_POLL_FREQ);
+    }
+
+    _waitEvent();
+}
+
+test('find SmartOS image', function _test(t) {
+    common.testFindSmartosImage(t, function _findSmartosCb(latest) {
+        smartosImageUUID = latest;
+        t.end();
+    });
+});
+
+test('load existing VMs', function _test(t) {
+    var opts = {};
+
+    opts.fields = ['uuid'];
+    opts.log = mocks.Logger;
+
+    vmadm.lookup({}, opts, function _onLookup(err, vms) {
+        t.ifError(err, 'vmadm lookup');
+        if (vms) {
+            vms.forEach(function _pushVm(vm) {
+                existingVms.push(vm.uuid);
+            });
+        }
+        t.end();
+    });
+});
+
+test('starting FsWatcher', function _test(t) {
+    function _onVmUpdate(vmUuid, updateType /* , updateObj */) {
+        // ignore events from VMs that existed when we started
+        if (existingVms.indexOf(vmUuid) === -1) {
+            events.push({
+                event: updateType,
+                timestamp: (new Date()).toISOString(),
+                vmUuid: vmUuid
+            });
+        }
+    }
+
+    watcher = new FsWatcher({
+        log: mocks.Logger,
+        updateVm: _onVmUpdate
+    });
+
+    t.ok(watcher, 'created FsWatcher');
+
+    watcher.start();
+
+    t.end();
+});
+
+test('create VM', function _test(t) {
+    var eventIdx = events.length;
+    var payload = {
+        alias: 'vm-agent_testvm',
+        brand: 'joyent-minimal',
+        image_uuid: smartosImageUUID,
+        quota: 10
+    };
+
+    payload.log = mocks.Logger;
+
+    vmadm.create(payload, function _vmadmCreateCb(err, info) {
+        t.ifError(err, 'create VM');
+        if (!err && info) {
+            t.ok(info.uuid, 'VM has uuid: ' + info.uuid);
+            smartosVmUUID = info.uuid;
+            waitEvent(t, 'create', smartosVmUUID, eventIdx);
```

Comment:
Is checking for `eventidx`useful? It seems we generate one single event after another sequentially, so eventIdx would always be `events.length`. Why no removing the event that was successfully checked by `waitEvent` each time and check only at index `0`? Or maybe even use a variable instead of a list?